crates/ironrdp-cliprdr/src/lib.rs

```diff
@@ -4,3 +4,197 @@
+        )
+    }
+
+    fn handle_error_tranistion(&mut self, err: ClipboardError) -> PduResult<Vec<SvcMessage>> {
+        // Failure of clipboard is not an critical error, but we should properly report it
+        // and transition channel to failed state.
+        self.state = CliprdrState::Failed;
+        error!("CLIPRDR(clipboard) failed: {err}");
+
+        Ok(vec![])
+    }
+
+    fn handle_server_capabilities(&mut self, server_capabilities: Capabilities) -> PduResult<Vec<SvcMessage>> {
+        self.capabilities.downgrade(&server_capabilities);
+        self.state = CliprdrState::Initialization(CliprdrInitializationState::ServerCapabilitiesReceived);
+
+        // Do not send anything, wait for monitor ready pdu
+        Ok(vec![])
+    }
+
+    fn handle_monitor_ready(&mut self) -> PduResult<Vec<SvcMessage>> {
+        let response = [
+            ClipboardPdu::Capabilites(self.capabilities.clone()),
+            ClipboardPdu::TemporaryDirectory(ClientTemporaryDirectory::new(self.temporary_directory.clone())?),
+            ClipboardPdu::FormatList(self.build_local_format_list()?),
+        ]
+        .into_iter()
+        .map(into_cliprdr_message)
+        .collect();
+
+        self.state = CliprdrState::Initialization(CliprdrInitializationState::WaitForServerFormatListResponse);
+        Ok(response)
+    }
+
+    fn handle_format_list_response(&mut self, response: FormatListResponse) -> PduResult<Vec<SvcMessage>> {
+        match self.state {
+            CliprdrState::Initialization(_) => match response {
+                FormatListResponse::Ok => {
+                    self.state = CliprdrState::Ready;
+                    info!("CLIPRDR(clipboard) virtual channel has been initialized");
+                }
+                FormatListResponse::Fail => {
+                    return self.handle_error_tranistion(ClipboardError::FormatListRejected);
+                }
+            },
+            CliprdrState::Ready => {
+                // TODO(@pacmancoder): Currently ignored as data transfer is not implemented,
+                // but in the next feature development interation we should handle this correctly
+            }
+            CliprdrState::Failed => unreachable!(),
+        }
+
+        Ok(vec![])
+    }
+}
+
+impl Default for Cliprdr {
+    fn default() -> Self {
+        Self {
+            state: CliprdrState::Initialization(CliprdrInitializationState::WaitForServer),
+            capabilities: Capabilities::new(
+                ClipboardProtocolVersion::V2,
+                ClipboardGeneralCapabilityFlags::USE_LONG_FORMAT_NAMES,
+            ),
+            temporary_directory: ".cliprdr".to_string(),
+        }
+    }
+}
+
+impl StaticVirtualChannel for Cliprdr {
+    fn channel_name(&self) -> ChannelName {
+        Self::CHANNEL_NAME
+    }
+
+    fn process(&mut self, payload: &[u8]) -> PduResult<Vec<SvcMessage>> {
+        let pdu = decode::<ClipboardPdu>(payload)?;
+
+        let messages = match &self.state {
```

Comment:
It's unclear to me what the advantage of maintaining this explicit state machine here is. Regardless of state, we are `match`ing on `pdu` and responding based on its type. AFAICT, this logic is only checking that the server gives us PDUs in the expected order -- however I think we can safely assume that the server is working appropriately, and simply respond based on what it's giving us.

Reply:
The advantage of this is that when we keep track of the state via a state machine, it gives us the opportunity to make error reporting more clean (on which exact communication stage we received unexpected/faulty PDU), it also allows to more logically split code into methods/types when code will grow upon adding different formats handling (initialization code is almost independent of an actual data transition logic). Also, it leaves less place for human error (the states will eventually store operation-specific fields which will be only accessible when matched on the correct state).

I think for such a complex thing as a clipboard it makes sense, but I may be wrong. @CBenoit what is your opinion on that?

Reply:
Though no state machine at all is perfectly viable and more readable for a small channel, I agree with you @pacmancoder 
If you think the CLIPRDR is going to grow into something way more complex that it currently is, then it makes sense to me. :+1:

Reply:
> it gives us the opportunity to make error reporting more clean (on which exact communication stage we received unexpected/faulty PDU)

I think this point still begs the question of whether tracking unexpected/faulty PDUs is actually a valid concern for the client. IMO a faulty server implementation isn't worth designing for.

> it also allows to more logically split code into methods/types when code will grow upon adding different formats handling (initialization code is almost independent of an actual data transition logic)

I'm not seeing how this is the case. From my perspective it seems that we can break down the internal methods/types used by `process` to whatever degree of resolution we want to in either approach.

> Also, it leaves less place for human error (the states will eventually store operation-specific fields which will be only accessible when matched on the correct state).

I suppose that's true at least theoretically. IMO, presuming my other critiques hold water, then this concern alone isn't large enough to justify all of the additional code needed for explicit state management.

I do notice that there's a `Sequence` trait which we `impl` for various state machine types that do something similar, e.g. 

https://github.com/Devolutions/IronRDP/blob/b9de8ba635ac0974ad792dd4170f3a94907a9e70/crates/ironrdp-connector/src/connection.rs#L262

I can see an argument for keeping consistency there, however in my view the "implicit state machine" version gets the job done just fine and is much easier to read / follow, e.g. 9c7e9daca02400b1a1e4f9c19d7a08d2677abc3d:

```rust
impl StaticVirtualChannel for Cliprdr {
    // ...

    fn process(&mut self, payload: &[u8]) -> PduResult<Vec<SvcMessage>> {
        let pdu = decode::<ClipboardPdu>(payload)?;

        match pdu {
            ClipboardPdu::Capabilites(caps) => self.handle_server_capabilities(caps),
            ClipboardPdu::MonitorReady => self.handle_monitor_ready(),
            ClipboardPdu::FormatListResponse(response) => self.handle_format_list_response(response),
            _ => todo!(),
        }
    }

    // ...
}
```

Reply:
Fair arguments.

> > Also, it leaves less place for human error (the states will eventually store operation-specific fields which will be only accessible when matched on the correct state).
>
> I suppose that's true at least theoretically. IMO, presuming my other critiques hold water, then this concern alone > isn't large enough to justify all of the additional code needed for explicit state management.

Operation-specific fields only accessible when matching on the correct state is still an interesting property, but it’s likely we can benefit from this without having a different state for absolutely each step in the sequence.

When it’s really needed, this approach could works as well:
```rust
#[derive(Default)]
enum CliprdrState {
    #[default]
    Idle,
    SpecialState { some_value: u64 },
}

impl Cliprdr {
    fn handle_start_special_thing(&mut self, context: SpecialContext) -> PduResult<Vec<SvcMessage>> {
        self.state = CliprdrState::SpecialState {
            special_value: context.value,
        };

        Ok(Vec::new())
    }

    fn handle_continue_special_thing(&mut self, other_context: SpecialOtherContext) -> PduResult<Vec<SvcMessage>> {
        let CliprdrState::SpecialState { special_value } = std::mem::take(&mut self.state) else { // consume the special state
            return Err(…);
        };

        // do something with `special_value` and `other_context`
    }
}

impl StaticVirtualChannel for Cliprdr {
    // ...

    fn process(&mut self, payload: &[u8]) -> PduResult<Vec<SvcMessage>> {
        let pdu = decode::<ClipboardPdu>(payload)?;

        match pdu {
            ClipboardPdu::Capabilites(caps) => self.handle_server_capabilities(caps),
            ClipboardPdu::MonitorReady => self.handle_monitor_ready(),
            ClipboardPdu::FormatListResponse(response) => self.handle_format_list_response(response),
            ClipboardPdu::StartSpecialThing(context) => self.handle_start_special_thing(context),
            ClipboardPdu::ContinueSpecialThing(other_context) => self.handle_continue_special_thing(other_context),
            _ => todo!(),
        }
    }

    // ...
}
```

It all depends on how much there is to keep track of imo.

Reply:
Ok, let's do this -- I'll remove the state machine, Currently for initialization we don't need it anyway. Later we will see if we need something to improve state management and refactor code appropriately.